Adding real-time status updates

Alexey-ebi opened the pull request.

Alexey-ebi committed:
Adding new asgi service

Alexey-ebi committed:
Addgin js accept ws messages and change text, color of status field

Alexey-ebi committed:
Rewriting consumers in async manner

Alexey-ebi committed:
Adding real-time submission status update upon data upload

Alexey-ebi committed:
Adding js code to disable and activate buttons

Alexey-ebi committed:
Adding status updates for validation step

Alexey-ebi committed:
Adding status updates to submission stage

bunop commented on django-data/image/biosample/tasks.py:
```diff
@@ -108,6 +110,9 @@ def on_failure(self, exc, task_id, args, kwargs, einfo):
             "Error in biosample submission: %s" % str(exc))
         submission_data.submission_obj.save()
 
+        asyncio.get_event_loop().run_until_complete(send_message_to_websocket(STATUSES.error.value[1],
```
Usually, I try to write python code using 80 characters column width. This seems frustrating, but have many advantages: You could easily revise code on GitHub web client or `git diff` and you will non see the `E501` linter warning with python IDE. In such way you can easily spot a more serious problem in coding. I worked on projects where you can't submit a PR if you have lines greater than 80 characters width (and others linter warnings). The only exception on such rule is a web link in comment, that should contain only the web link. You can find a lot of examples in my code. This is how I will write such line:
```python
        asyncio.get_event_loop().run_until_complete(
            send_message_to_websocket(
                STATUSES.error.value[1],
                submission_data.submission_id
            )
        )
```
When I have a lot of conditions wich increase the indentation depth, I use to write an helper function to make coding easier

Alexey-ebi replied:
Fixed

bunop commented on django-data/image/biosample/tasks.py:
```diff
@@ -164,6 +169,9 @@ def submit(self, submission_data):
             submission_data.submission_obj.message = message
             submission_data.submission_obj.save()
 
+            asyncio.get_event_loop().run_until_complete(send_message_to_websocket(STATUSES.ready.value[1],
```
Will be `submission_data.submission_obj.get_status_display()` more readable that `STATUSES.ready.value[1]`? Or will adding a method in `common.constants.EnumMixin` as:
```python
class EnumMixin():
    """Common methods for my Enum classes"""

    @classmethod
    def get_value(cls, member):
        """Get numerical representation of an Enum object"""

        return cls[member].value[0]

    @classmethod
    def get_value_display(cls, value):
        for el in cls:
            if el.value[0] == value:
                return el.value[1]

        return Exception("value %s not in %s" % (value, cls))
```
And then, get message as:
```python
STATUSES.get_value_display(READY)
```

Alexey-ebi replied:
Agree, added new method to EnumMixin

bunop commented on django-data/image/common/helpers.py:
```diff
@@ -72,3 +74,15 @@ def format_callback(obj):
+async def send_message_to_websocket(message, pk):
+    """
+    Function will create websocket object and send message to django-channels
+    :param message: message to send to websocket
```
You need to add an empty line in order to correct render this help with sphinx:
```python
async def send_message_to_websocket(message, pk):
    """
    Function will create websocket object and send message to django-channels

    :param message: message to send to websocket
    :param pk: primary key of submission
    """
```
 Or you could use the the [google style](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html#overview) since we are importing the `sphinx.ext.napoleon` externsion

Alexey-ebi replied:
Used google style

bunop commented on django-data/image/biosample/tasks.py:
```diff
@@ -193,6 +201,9 @@ def submit(self, submission_data):
             submission_data.submission_obj.message = message
             submission_data.submission_obj.save()
 
+            asyncio.get_event_loop().run_until_complete(send_message_to_websocket(STATUSES.ready.value[1],
```
Same `E501` warning as before and valid for all `>80` characters lines

Alexey-ebi replied:
Fixed

bunop commented on docker-compose.yml:
```diff
@@ -63,6 +63,35 @@ services:
+        source: ./django-data/
+        target: /var/uwsgi/
+
+      - type: bind
```
Maybe we can remove this bind: it is needed by `uwsgi` to display InjectTool version and git link at the bottom of the page using `image_app.template_tags.git_describe` and `image_app.template_tags.get_git_link`

Alexey-ebi replied:
Agree, done.

bunop commented on django-data/image/submissions/templates/submissions/submission_detail.html:
```diff
@@ -120,3 +120,73 @@ <h1>Submission Detail for <em>{{ submission.title }}</em></h1>
+    if (current_status !== 'Ready') {
+      submit_button.classList.add('disabled');
+    } else {
+      submit_button.classList.remove('disabled');
```
Need also to remove `disabled` from `edit_data`, `validate` and `danger_zone`. When status is `READY`. Please see `submission.templatetags.submission_tags` function and function called in `image_app.models.Submission`

Alexey-ebi replied:
Agree, fixed

bunop commented on django-data/image/submissions/templates/submissions/submission_detail.html:
```diff
@@ -120,3 +120,73 @@ <h1>Submission Detail for <em>{{ submission.title }}</em></h1>
+    var submit_button = document.querySelector('#submit-button');
+    if (current_status === 'Waiting' || current_status === 'Submitted') {
+      edit_data.classList.add('disabled');
+      danger_zone.classList.add('disabled');
```
also `submit_button` need to be disabled. Is this already disabled?

Alexey-ebi replied:
Yes, it will be disables in third condition 
If (current_status !== 'Ready')

bunop commented on django-data/image/submissions/templates/submissions/submission_detail.html:
```diff
@@ -120,3 +120,73 @@ <h1>Submission Detail for <em>{{ submission.title }}</em></h1>
+  current_status = document.querySelector('#status-log').textContent;
+  document.querySelector('#status-log').className = getColor(current_status);
+  });
+</script>
```
There's also the submission message defined in `submissions.views.MessagesSubmissionMixin` which changes according `django.contrib.messages` and its related `MIDDLEWARE` and `context_processors` modules. Should those messages pass using websocket? will page reload simplify dealing with messages in templates and buttons check?

Alexey-ebi replied:
Is it related to notifications at the top of the page? If yes, I think we should create another Issue with this task...

bunop commented on django-data/image/common/helpers.py:
```diff
@@ -72,3 +74,15 @@ def format_callback(obj):
+    :param pk: primary key of submission
+    """
+    # Need to have it here as in case with small test data message sent to websocket will overcome response from server
+    time.sleep(3)
```
Will this increase time need for testing? This function need to be mocked during tests

Alexey-ebi replied:
Wrote tests for submissions_ws itself, next thing is to fix existing unittests to use mock instead of real function

bunop replied:
Ty. Tomorrow I could give you AN example

bunop replied:
Hi Alexey, 

there are two ways to mocking object in your case. I'm not an expert in mocking channels, but if you take [test_task.py](https://github.com/cnr-ibba/IMAGE-InjectTool/blob/b10ceb20b08eb5009cd29af59265900c3a162508/django-data/image/biosample/tests/test_tasks.py#L107), you can see an example of I mock object using patch decorator. This is patched only in this test. Outside of this, the function will called normally as the site does. Another way to mocking object, could be done like this
```python
class SubmitTestCase(SubmitMixin, RedisMixin, TestCase):

    @classmethod
    def setUpClass(cls):
        # calling my base class setup
        super().setUpClass()

        # starting mocked objects
        cls.mock_ws_patcher = patch(
            'asyncio.get_event_loop')
        cls.mock_ws = cls.mock_ws_patcher.start()

    @classmethod
    def tearDownClass(cls):
        # stopping mocked object
        cls.mock_ws_patcher.stop()

        # calling base method
        super().tearDownClass()

    def setUp(self):
        # call Mixin method
        super().setUp()

        # setting tasks
        self.my_task = SubmitTask()
```
With this code, you can execute `docker-compose run --rm uwsgi pytest --verbosity=2 biosample/tests/test_tasks.py::SubmitTestCase` without using redis and channel (check asgi log with docker logs)
By adding the `setUpClass` and `tearDownClass`, you will mock asyncio for each test inside you function. The only drawback is that since setUpClass is called once when generating test, you can't distinguish when and where you mocked function is called, for example with `self.assertTrue(self.mock_ws.called)`
If you define the same function in `setUp` as defined [here](https://github.com/cnr-ibba/IMAGE-InjectTool/blob/b10ceb20b08eb5009cd29af59265900c3a162508/django-data/image/biosample/tests/common.py#L66), you will setUp the mock object before each method is called, so you can distinguish when you have called your methods.

Alexey-ebi replied:
Done

bunop commented on django-data/image/image/settings.py:
```diff
@@ -63,6 +64,7 @@
     'animals',
     'samples',
     'crbanim',
+    'submissions_ws'
```
Should this name remind real time status update?

Alexey-ebi replied:
"ws" should remind that we will work with websockets protocol here. For submissions we will also have:
 - real-time updates of number of animal and samples in submission
 - real-time updates of validation results
So I don't think that we should rename this app

bunop reviewed (commented):
Many thanks, Alexey, for your work. There are some points I would like to discuss, they are provided as code review comments. A few point more:
- websocket are called when performing tests: you can see it in docker-compose logs while you are running test which implies status update (every time you write `asyncio.get_event_loop()....`). Moreover, tests are run by uploading fixtures on a test database, but channels works with real submission id, so you can see status changing for submission 1 while test are performed on background. Will be possible to mock up `send_message_to_websocket` in order to avoid sending messages during tests?
- I don't have experience in django channels, but it will be possible to test them using unittest? will this be difficult or time consuming?
- Would you like to add a sphinx document regarding real time status updates?

Alexey-ebi committed:
Delete unnecessary binding

Alexey-ebi committed:
Adding get_value_display method to EnumMinx class

Alexey-ebi committed:
Fixing lint problems

Alexey-ebi committed:
Removing disabled class from appropriate buttons when status is Ready

Alexey-ebi committed:
Adding tests for ws consumer

bunop commented:
This will close #14

Alexey-ebi committed:
Fixing tests for cryoweb and crbanim

Alexey-ebi committed:
Fixing tests for validation

Alexey-ebi committed:
Fixing tests for biosample submission

bunop committed:
:white_check_mark: mocking channels in cryoweb test_import_has_data

Added last mock object to channels. No more call of asgi during tests

bunop committed:
:twisted_rightwards_arrows: Merge branch 'master' into adding_channels

Conflicts:
	django-data/image/validation/tasks.py
	django-data/image/validation/tests/test_tasks.py

Conflicts were resolved. image_validation library upgraded,
remember to re-build images with docker-build. You need also
to apply a new migration from master branch

bunop commented:
Ty @Alexey-ebi for last fixes. I'm resolving conflicts and adding a mock asyncio you have missed. What do you prefer? You are currently working on this branch? Could I push directly on your branch or should send a PR to your branch?

Alexey-ebi commented:
Hi Paolo, you could push directly to this branch and merge it. I started to work on: 
1. real-time notifications on top of the page
2. real-time updates of number of animals and samples in submission
3. real-time updates of validation results
All these will be in another branch and issue.

bunop commented:
Ok. I resolved some conflicts with master. You have to re-build containers and apply migrations since I updated `image_validation` and I fix a thing on `image_app.models`. Alexey, would you like to add a simple page in sphinx describing briefly how channel works? something like [general documentation](https://www.wp5image.eu/image/docs/index.html#general-documentation) section? I could help you to setup this page, if you need help to working with sphinx

Alexey-ebi commented:
I've never worked with sphinx so it could take some time... But I think that we have more urgent tasks to do and documentation could be added later when all critical things will be done.

Alexey-ebi commented:
I could add it when finish to work on all real-time updates tasks.

bunop commented:
It's not so difficult. It consist to write *restructured text* documents with some keys which could reference to other documentation section inside or outside the project. Take a look at [this](https://raw.githubusercontent.com/cnr-ibba/IMAGE-InjectTool/master/django-data/image/docs/source/general/asynchronous_tasks.rst) document and [how](https://www.wp5image.eu/image/docs/general/asynchronous_tasks.html) it is rendered. Ok to pospone those docs but I think that they are important since this tool is becoming more and more complex. Outside our developing efforts, could be used also to describe how work is done.

bunop commented:
So I will merge this PR. Thank you again for your contribute!
